### DISCUSSION TITLE
server side socket.data properties are cloned?

### DISCUSSION THREAD

**Discussion by nedgrady** [socketio/socket.io#4935] [Q&A] [answered]
Issue/observation/confusion: It looks like socket.data.xxx is somehow re-assigned to a different object with the same properties, and I'm very confused.

I have middleware to attach an 'active user' to a given socket instance. Here is a simplified version that assumes there is only ever one user.

```
// identifyAllSocketsAsTheSameUser.ts
const singleActiveUser = {
	uniqueIdentifier: "Single user from identifyAllSocketsAsTheSameUser",
	connections: new Set<TicTacWoahServerSocket>(),
}

const handleToSingleActiveUser = {
	activeUser: singleActiveUser,
}

export const identifyAllSocketsAsTheSameUser: TicTacWoahSocketServerMiddleware = (socket, next) => {
	singleActiveUser.connections.add(socket)
	socket.data.activeUser = singleActiveUser
	socket.data.handle = handleToSingleActiveUser

	next()
}
```

Now create the server & use middleware:

```
// index.ts
import express from "express"
import { createServer } from "http"

const app = express()
const httpServer = createServer(app)

const io: TicTacWoahSocketServer = new Server(httpServer, {
	cors: {
		origin: ["https://admin.socket.io", "http://localhost:5173"],
		methods: ["GET", "POST"],
		credentials: true,
	},
})

io.use(identifyAllSocketsAsTheSameUser)

// add info endpoint to easily view internal details
app.get("/info", async (_request, response) => {

	const allTheSockets: TicTacWoahServerSocket[] = (await io.fetchSockets())

	const con1 = allTheSockets[0]
	const con2 = allTheSockets[1]

	response.json({
		areSameData: con1.data.activeUser === con2.data.activeUser,
		areSameHandle: con1.data.handle.activeUser === con2.data.handle.activeUser,
	})
})


httpServer.listen(8080)
```

Now connect with two clients and hit the info endpoint, and the response is as follows:

```
{"areSameData":false, "areSameHandle":true}
```

I would have expected both object references to remain pointing to the original activeUser/activeUserHandle as defined in the middleware (and thus expected  `{"areSameData":true, "areSameHandle":true}` ), but it looks like the `socket.data.activeUser` property is somehow being re-assigned to a different object between the two connections?

Is this intended behaviour, and if so is it mentioned anywhere in the docs? Or perhaps I've misunderstood something? Any thoughts much appreciated.

**Comment by darrachequesne** [comment, marked as answer]
Hi! I see two possible explanations:

- you have multiple Socket.IO servers, then in that case `io.fetchSockets()` returns all the connections in the cluster
- you use the admin UI, which [tracks every change](https://github.com/socketio/socket.io-admin-ui/blob/158864989dddeba67df9975a4cad48ef310f8c80/lib/index.ts#L342-L347) to the `data` attribute

Could you please check?

**Reply by nedgrady** [reply]
[Yes the server in question is using the admin ui](https://github.com/nedgrady/tic-tac-woah/blob/master/packages/server/src/index.ts)

This explains why I couldn't recreate the same behaviour in my unit tests as these don't use admin UI.

So perhaps using object reference equality for objects attached to socket.data isn't the best way to go?

**Reply by darrachequesne** [reply]
> So perhaps using object reference equality for objects attached to socket.data isn't the best way to go?

Yes indeed!

**Reply by nedgrady** [reply]
Thanks v much for the quick reply - is this gotcha something that might want to be mentioned in the docs (if it isn't already). Would be very happy to knock up a quick note myself if so

**Reply by darrachequesne** [reply]
Sure, please do! The source of the website is here: https://github.com/socketio/socket.io-website
